db/design_doc.go

```diff
@@ -265,3 +287,396 @@ func stripSyncProperty(row *sgbucket.ViewRow) {
+			ViewAllDocs:    sgbucket.ViewDef{Map: alldocs_map, Reduce: "_count"},
+			ViewImport:     sgbucket.ViewDef{Map: import_map, Reduce: "_count"},
+			ViewOldRevs:    sgbucket.ViewDef{Map: oldrevs_map, Reduce: "_count"},
+			ViewSessions:   sgbucket.ViewDef{Map: sessions_map},
+			ViewTombstones: sgbucket.ViewDef{Map: tombstones_map},
+		},
+		Options: &sgbucket.DesignDocOptions{
+			IndexXattrOnTombstones: true, // For ViewTombstones
+		},
+	}
+
+	sleeper := base.CreateDoublingSleeperFunc(
+		11, //MaxNumRetries approx 10 seconds total retry duration
+		5,  //InitialRetrySleepTimeMS
+	)
+
+	// add all design docs from map into bucket
+	for designDocName, designDoc := range designDocMap {
+
+		//start a retry loop to put design document backing off double the delay each time
+		worker := func() (shouldRetry bool, err error, value interface{}) {
+			err = bucket.PutDDoc(designDocName, designDoc)
+			if err != nil {
+				base.Warn("Error installing Couchbase design doc: %v", err)
+			}
+			return err != nil, err, nil
+		}
+
+		description := fmt.Sprintf("Attempt to install Couchbase design doc bucket : %v", designDocName)
+		err, _ := base.RetryLoop(description, worker, sleeper)
+
+		if err != nil {
+			return pkgerrors.Wrapf(err, "Error installing Couchbase Design doc: %v", designDocName)
+		}
+	}
+
+	base.Logf("Design docs successfully created for view version %s.", DesignDocVersion)
+
+	return nil
+}
+
+// Issue a stale=false queries against critical views to guarantee indexing is complete and views are ready
+func WaitForViews(bucket base.Bucket) error {
+	viewErrors := make(chan error, 3)
+	var viewsWg sync.WaitGroup
+	views := []string{ViewChannels, ViewAccess, ViewRoleAccess}
+
+	base.Logf("Verifying view availability for bucket %s...", bucket.GetName())
+
+	for _, viewName := range views {
+		viewsWg.Add(1)
+		go func(view string) {
+			defer viewsWg.Done()
+			viewErr := waitForViewIndexing(bucket, DesignDocSyncGateway(), view)
+			if viewErr != nil {
+				viewErrors <- viewErr
+			}
+		}(viewName)
+	}
+
+	viewsWg.Wait()
+	if len(viewErrors) > 0 {
+		err := <-viewErrors
+		close(viewErrors)
+		return err
+	}
+
+	base.Logf("Views ready for bucket %s.", bucket.GetName())
+	return nil
+
+}
+
+// Issues stale=false view queries to determine when view indexing is complete.  Retries on timeout
+func waitForViewIndexing(bucket base.Bucket, ddocName string, viewName string) error {
+	var vres interface{}
+	opts := map[string]interface{}{"stale": false, "key": fmt.Sprintf("view_%d_ready_check", viewName), "limit": 1}
+	for {
+		err := bucket.ViewCustom(ddocName, viewName, opts, &vres)
+		// Retry on timeout error, otherwise return
+		if err == nil || err != base.ErrViewTimeoutError {
+			return err
+		} else {
+			base.Logf("Timeout waiting for view %q to be ready for bucket %q - retrying...", viewName, bucket.GetName())
+		}
+	}
+
+}
+
+func removeObsoleteDesignDocs(bucket base.Bucket, previewOnly bool) (removedDesignDocs []string, err error) {
```

Comment:
`removeObsoleteDesignDocs()` could be  split into two functions:

```
func getObsoleteDesignDocs(bucket base.Bucket) (obsoluteDesignDocs []string, err error)
```

and 

```
func removeObsoleteDesignDocs(obsoluteDesignDocs []string, previewOnly bool) (removedDesignDocs []string, err error)
```

where `getObsoleteDesignDocs()` should be possible to write a self-contained unit test against

Reply:
I think we've got already good coverage of the 'get' functionality as part of the existing TestRemoveObsoleteDesignDocs, so I don't think this is required for unit testing.

I prefer the code as-is - it means we don't duplicate the loop processing and ddoc name calculation, and we only need to modify once place if we want to change something.